### DISCUSSION TITLE
How to construct PEO melts at atomistic level

### DISCUSSION THREAD

**Discussion by YanChen32** [marrink-lab/polyply_1.0#320] [General]
Dear all,

I am currently working on simulating a mixed melt of PEO and PVA using the OPLSAA force field. I have already completed all the tutorials provided by polyply_1.0. However, most of the examples in polyply_1.0 are currently focused on coarse-grained models. Therefore, I have a few questions regarding how to construct all-atom models using the OPLSAA force field.

Firstly, I would like to know the process of building all-atom models based on the OPLSAA force field in polyply_1.0. Is there a comprehensive example available that demonstrates the usage of OPLSAA in polyply_1.0?

Additionally, I would like to understand how to create a melt model from scratch using a customized OPLSAA force field file, e.g., PEO or PVA. Specifically, I am interested in the steps and procedures involved in this process.

Many thanks,

Yan

**Comment by fgrunewald** [comment]
Dear Yan, 

Thanks for opening this discussion! We'll try our best to guide you through the process. Unfortunately, there is no comprehensive example yet. However, we've been working with OPLS quite a bit (especially @ricalessandri is the expert here).  Let's start:

# Creating a polymer melt with OPLS FF

## 1 - generating an itp file
Creating an initial simulation setup with OPLS is fairly easy and involves two steps. First you need the itp file of the polymer and secondly you need initial system coordinates. If you want to do this for PEG for example you need the following command to make the itp file, which follows the same idea s for CG systems. 
```
polyply gen_params -lib oplsaaLigParGen -seq OHter:1 PEO:100 OHter:1 -o PEO.itp -name PEO
```
This command will generate an itp file for 100 monomers of PEO using the OPLS LigParGen force-field. The LigParGen only means that the charges and atom-type assignment were done using LigParGen. We'll come to that in a minute. Note that you have to use the `OHter:1` residue in order to terminate the PEO chain with an alcohol group. In your particular case we can later change those to CH3 if it's more suitable. 

## 2 - building the initial melt 
To generate a melt you need a topology file of the system. In the PEG case the non-bonded parameters are provided by the LigParGen server, so the following topology file will do. Note that we aim to generate 100 polymers of PEO. 
```
#include "oplsaa.ff/forcefield.itp"
#include "PEO.itp"

[ system ]
test
[ molecules ]
PEO 100
```
Next you simply generate the melt at a particular density using the following command. This should not take very long, but depends a little bit on your python version and machine. For me it took 30s. 
```
 polyply gen_coords -p system.top -o start.gro -dens 1000 -name melt
```

**NOTE**: @ricalessandri made an edit here given that, based on [this recent PR](https://github.com/marrink-lab/polyply_1.0/pull/317), the required `.top` file is simpler.

## 3 - preparing to run a simulation
Part of the structure generation is an energy minization. When you look at the start.gro file you will see that the PEG looks like the bonds are not correct and the structure is bad. However, polyply scales the residue coordinates and they first have to "unflod" in an energy minimization. So next you run an energy minization using the default OPLS settings. Make sure however to do the energy minimization with flexible bonds not constraints. After that you can start the melt simulation in NVT/NpT as you want. 

# Creating parameters for a custom polymer
In order to create itp-files for a custom polymer, you need to define the bonded parameters, charges and atom-types for the repeat unit as well as how they change when they are linked together. All this information is stored in a .ff file. Luckily for you for OPLS @ricalessandri  has created a [small script](https://github.com/ricalessandri/itp_to_ff) which allows you to generate the .ff input files  you need from an itp file of a tetramer for example. So typically we generate a tetramer using LigParGen server, then apply the script itp_to_ff.py, and finally you can generate the itp file as follows:
```
polyply gen_params -f *.ff   ...
````

## Watch out for tacticity
Polyply does not define tacticity, instead the tacticity has to be defined via improper dihedral angles. You do this by adding the following entry to the ff-files. Let us suppose for PVA the main backbone atoms in the monomer are called `CC1` CC2` and you have the `HCC2` and `O`  as the names of the hydrogen connected to the backbone carbon, where also the O group is attached. In other words the PVA repeat unit looks as follows:
```
          HCC2
          |
 - CC1 - CC2 -
          |
          O - H          
```
In this case you add the following two entries to any of the .ff files:
```
[ link ]
resname PVA
[ impropers ]
CC2 {"chiral": "S"} CC1 {"chiral": "S"} +CC1 {"chiral": "R|S"} O {"chiral": "S"} 2 -3.5264390e+01  3.3484625e+02 {"comment":"center S", "group":"link"}
[ edges ]
CC2  CC1
CC2 O
CC2 +CC1

[ link ]
resname "PVA"
[ impropers ]
CC2 {"chiral": "R"} CC1 {"chiral": "R"} +CC1 {"chiral": "S|R"} O {"chiral": "R"} 2 3.5264390e+01  3.3484625e+02 {"comment":"center R", "group":"link"}
[ edges ]
CC2  CC1
CC2 O
CC2 +CC1
```
Next to generate a polymer with random tacticity you have to define the chirality of each monomer randomly. To do so call polyply gen_seq as shown below. 
```
polyply gen_seq -from_string A:100:1:PVA-1.0 -o PVA.json -name PVA -seq A -label 0:chiral:R-0.5,S-0.5
```
To generate iso-tactic PVA you use this command:
```
-label 0:chiral:R-1.0
```
Note that the last part `-label 0:chiral:R-1.0` defines the percentage of monomers that have R or S chirality. Any float combination is valid. If you have more special patterns like triads that can also be included, but we can discuss when that is needed. The itp file now has to be generated using this sequence file:
```
polyply gen_params -f <your_ff_files> -seqf PVA.json -o PVA.itp -name PVA
```
That's it!

If you have any question or assistance let me know. In case you want to submit your PVA/PEG to the polyply library and make it available to others, we're happy to assist. And ultimately if you like the package cite us and give us a star on GitHub.

**Reply by YanChen32** [reply]
Thank you very much for your prompt and detailed response. I sincerely appreciate it. I'm more than happy to submit the well-tuned PVA/PEG OPLSAA force field to the Polyply library. 
Following your guidance, I have successfully utilized the OPLSAA force field for MD simulations of PEO polymers. However, I encountered difficulties in developing the force field for PVA. The challenge arose when attempting to generate the topology file for the PVA monomer using LigParGen. Directly submitting the monomer (-CH2-CH(OH)-) resulted in an obviously incorrect topology file. Furthermore, my attempts to submit a longer chain (5 units with termini) have been unsuccessful thus far. Could you kindly provide guidance on effectively submitting the monomer file for polymers to LigParGen? I would greatly appreciate your assistance.

**Reply by ricalessandri** [reply]
Sometimes the LigParGen server misbehaves, i.e., sometimes submitting a day after simply works. Nevertheless, I should have a still-functional local installation of it, so I'll try that too for PVA on Monday. And follow up in this thread. Please @YanChen32 ping me if you don't hear from me in a few days.

**Reply by YanChen32** [reply]
Hi @ricalessandri, I have successfully generated the ITP file for PVA. I noticed that the atomic charges only vary slightly for PVA with different unit numbers.

For instance, in the case of [H]OCC(O)CC(O)CCO[H], the [atoms] part in ITP file reads

[ atoms ]
;   nr       type  resnr residue  atom   cgnr     charge       mass  
     1   opls_800      1    UNK   H00      1     0.4168     1.0080 
     2   opls_801      1    UNK   O01      1    -0.7089    15.9990 
     3   opls_802      1    UNK   C02      1      0.063    12.0110 
     4   opls_803      1    UNK   C03      1     0.1782    12.0110 
     5   opls_804      1    UNK   O04      1    -0.6893    15.9990 
     6   opls_805      1    UNK   C05      1    -0.2459    12.0110 
     7   opls_806      1    UNK   C06      1     0.1759    12.0110 
     8   opls_807      1    UNK   O07      1    -0.6888    15.9990 
     9   opls_808      1    UNK   C08      1    -0.2242    12.0110 
    10   opls_809      1    UNK   C09      1     0.1189    12.0110 
    11   opls_810      1    UNK   O0A      1    -0.6926    15.9990 
    12   opls_811      1    UNK   H0B      1     0.4113     1.0080 
    13   opls_812      1    UNK   H0C      1     0.0918     1.0080 
    14   opls_813      1    UNK   H0D      1     0.0918     1.0080 
    15   opls_814      1    UNK   H0E      1     0.1453     1.0080 
    16   opls_815      1    UNK   H0F      1     0.4217     1.0080 
    17   opls_816      1    UNK   H0G      1     0.1114     1.0080 
    18   opls_817      1    UNK   H0H      1     0.1114     1.0080 
    19   opls_818      1    UNK   H0I      1     0.1297     1.0080 
    20   opls_819      1    UNK   H0J      1     0.4103     1.0080 
    21   opls_820      1    UNK   H0K      1     0.1132     1.0080 
    22   opls_821      1    UNK   H0M      1     0.1132     1.0080 
    23   opls_822      1    UNK   H0N      1     0.0728     1.0080 
    24   opls_823      1    UNK   H0O      1     0.0728     1.0080 

In the case of [H]OCC(O)CC(O)CC(O)CC(O)CC(O)CC(O)CC(O)CC(O)CC(O)CC(O)CCO[H], the [ atoms ] part in ITP file becomes
[ atoms ]
;   nr       type  resnr residue  atom   cgnr     charge       mass  
     1   opls_800      1    UNK   H00      1     0.4163     1.0080 
     2   opls_801      1    UNK   O01      1    -0.7035    15.9990 
     3   opls_802      1    UNK   C02      1      0.067    12.0110 
     4   opls_803      1    UNK   C03      1      0.176    12.0110 
     5   opls_804      1    UNK   O04      1    -0.6895    15.9990 
     6   opls_805      1    UNK   C05      1    -0.2446    12.0110 
     7   opls_806      1    UNK   C06      1     0.1783    12.0110 
     8   opls_807      1    UNK   O07      1    -0.6949    15.9990 
     9   opls_808      1    UNK   C08      1    -0.2532    12.0110 
    10   opls_809      1    UNK   C09      1     0.1806    12.0110 
    11   opls_810      1    UNK   O0A      1    -0.6861    15.9990 
    12   opls_811      1    UNK   C0B      1    -0.2582    12.0110 
    13   opls_812      1    UNK   C0C      1     0.1788    12.0110 
    14   opls_813      1    UNK   O0D      1    -0.6856    15.9990 
    15   opls_814      1    UNK   C0E      1    -0.2582    12.0110 
    16   opls_815      1    UNK   C0F      1     0.1775    12.0110 
    17   opls_816      1    UNK   O0G      1    -0.6877    15.9990 
    18   opls_817      1    UNK   C0H      1    -0.2564    12.0110 
    19   opls_818      1    UNK   C0I      1     0.1769    12.0110 
    20   opls_819      1    UNK   O0J      1    -0.6876    15.9990 
    21   opls_820      1    UNK   C0K      1    -0.2575    12.0110 
    22   opls_821      1    UNK   C0M      1     0.1755    12.0110 
    23   opls_822      1    UNK   O0N      1    -0.6852    15.9990 
    24   opls_823      1    UNK   C0O      1    -0.2602    12.0110 
    25   opls_824      1    UNK   C0P      1     0.1607    12.0110 
    26   opls_825      1    UNK   O0Q      1    -0.6995    15.9990 
    27   opls_826      1    UNK   C0R      1     -0.261    12.0110 
    28   opls_827      1    UNK   C0S      1     0.1872    12.0110 
    29   opls_828      1    UNK   O0T      1    -0.7122    15.9990 
    30   opls_829      1    UNK   C0U      1    -0.2269    12.0110 
    31   opls_830      1    UNK   C0V      1     0.1763    12.0110 
    32   opls_831      1    UNK   O0W      1    -0.6927    15.9990 
    33   opls_832      1    UNK   C0X      2    -0.2104    12.0110 
    34   opls_833      1    UNK   C0Y      2     0.1153    12.0110 
    35   opls_834      1    UNK   O0Z      2    -0.6952    15.9990 
    36   opls_835      1    UNK   H10      2     0.4115     1.0080 
    37   opls_836      1    UNK   H11      2      0.089     1.0080 
    38   opls_837      1    UNK   H12      2      0.089     1.0080 
    39   opls_838      1    UNK   H13      2     0.1446     1.0080 
    40   opls_839      1    UNK   H14      2     0.4209     1.0080 
    41   opls_840      1    UNK   H15      2     0.1128     1.0080 
    42   opls_841      1    UNK   H16      2     0.1128     1.0080 
    43   opls_842      1    UNK   H17      2     0.1401     1.0080 
    44   opls_843      1    UNK   H18      2     0.4071     1.0080 
    45   opls_844      1    UNK   H19      2     0.1044     1.0080 
    46   opls_845      1    UNK   H1A      2     0.1044     1.0080 
    47   opls_846      1    UNK   H1B      2     0.1473     1.0080 
    48   opls_847      1    UNK   H1C      2     0.4048     1.0080 
    49   opls_848      1    UNK   H1D      2     0.1069     1.0080 
    50   opls_849      1    UNK   H1E      2     0.1069     1.0080 
    51   opls_850      1    UNK   H1F      2     0.1409     1.0080 
    52   opls_851      1    UNK   H1G      2     0.4088     1.0080 
    53   opls_852      1    UNK   H1H      2     0.1092     1.0080 
    54   opls_853      1    UNK   H1I      2     0.1092     1.0080 
    55   opls_854      1    UNK   H1J      2     0.1401     1.0080 
    56   opls_855      1    UNK   H1K      2     0.4083     1.0080 
    57   opls_856      1    UNK   H1M      2     0.1098     1.0080 
    58   opls_857      1    UNK   H1N      2     0.1098     1.0080 
    59   opls_858      1    UNK   H1O      2     0.1393     1.0080 
    60   opls_859      1    UNK   H1P      2     0.4098     1.0080 
    61   opls_860      1    UNK   H1Q      2       0.11     1.0080 
    62   opls_861      1    UNK   H1R      2       0.11     1.0080 
    63   opls_862      1    UNK   H1S      2     0.1353     1.0080 
    64   opls_863      1    UNK   H1T      2     0.4109     1.0080 
    65   opls_864      1    UNK   H1U      2     0.1135     1.0080 
    66   opls_865      1    UNK   H1V      3     0.1135     1.0080 
    67   opls_866      1    UNK   H1W      3     0.1341     1.0080 
    68   opls_867      1    UNK   H1X      3       0.42     1.0080 
    69   opls_868      1    UNK   H1Y      3     0.1183     1.0080 
    70   opls_869      1    UNK   H1Z      3     0.1183     1.0080 
    71   opls_870      1    UNK   H20      3     0.0913     1.0080 
    72   opls_871      1    UNK   H21      3      0.438     1.0080 
    73   opls_872      1    UNK   H22      3     0.1125     1.0080 
    74   opls_873      1    UNK   H23      3     0.1125     1.0080 
    75   opls_874      1    UNK   H24      3     0.1425     1.0080 
    76   opls_875      1    UNK   H25      3     0.4074     1.0080 
    77   opls_876      1    UNK   H26      3     0.1099     1.0080 
    78   opls_877      1    UNK   H27      3     0.1099     1.0080 
    79   opls_878      1    UNK   H28      3      0.072     1.0080 
    80   opls_879      1    UNK   H29      3      0.072     1.0080 

We can observe that the charges for PVA with different lengths appear similar. However, it is worth noting that the charges for the terminal groups in PVA and PEO are different. Therefore, it may be necessary to create separate terminal files for PEO and PVA, right? 

Then, before generating the .ff input files using the provided script (By the way, I just want to mention that I'm unable to open the link.), should we first extract the specific bond, angle, and dihedral angle terms for both PVA and its terminal groups and rebuild the ITP files, right?

Thank you so much for your time and assistance!

**Reply by YanChen32** [reply]
For example, I extract the ITP file for PVA as 
`;
; GENERATED BY LigParGen Server
; Jorgensen Lab @ Yale University 
;
[ atomtypes ]
  opls_136  C136    12.0110     0.000    A    3.50000E-01   2.76144E-01
  opls_158  C158    12.0110     0.000    A    3.50000E-01   2.76144E-01
  opls_154  O154    15.9990     0.000    A    3.12000E-01   7.11280E-01
  opls_140  H140     1.0080     0.000    A    2.50000E-01   1.25520E-01
  opls_140  H140     1.0080     0.000    A    2.50000E-01   1.25520E-01
  opls_140  H140     1.0080     0.000    A    2.50000E-01   1.25520E-01
  opls_155  H155     1.0080     0.000    A    0.00000E+00   0.00000E+00
[ moleculetype ]
; Name               nrexcl
PVA                   3
[ atoms ]
;   nr       type  resnr residue  atom   cgnr     charge       mass  
     1   opls_136      1    PVA   C01      1    -0.2459    12.0110 
     2   opls_158      1    PVA   C02      1     0.1759    12.0110 
     3   opls_154      1    PVA   O03      1    -0.6888    15.9990 
     4   opls_140      1    PVA   H04      1     0.1114     1.0080 
     5   opls_140      1    PVA   H05      1     0.1114     1.0080 
     6   opls_140      1    PVA   H06      1     0.1297     1.0080 
     7   opls_155      1    PVA   H07      1     0.4103     1.0080 
[ bonds ]
    2     1     1      0.1529 224262.400
    3     2     1      0.1410 267776.000
    4     1     1      0.1090 284512.000
    5     1     1      0.1090 284512.000
    6     2     1      0.1090 284512.000
    7     3     1      0.0945 462750.400

[ angles ]
;  ai    aj    ak funct            c0            c1            c2            c3 
    1     2     3     1    109.500    418.400
    1     2     6     1    110.700    313.800
    2     3     7     1    108.500    460.240
    2     1     4     1    110.700    313.800
    2     1     5     1    110.700    313.800
    4     1     5     1    107.800    276.144
    3     2     6     1    109.500    292.880

[ dihedrals ]
; IMPROPER DIHEDRAL ANGLES 
;  ai    aj    ak    al funct            c0            c1            c2            c3            c4            c5

[ dihedrals ]
; PROPER DIHEDRAL ANGLES
;  ai    aj    ak    al funct            c0            c1            c2            c3            c4            c5
    6    2    1    4        3       0.628   1.883   0.000  -2.510  -0.000   0.000
    6    2    1    5        3       0.628   1.883   0.000  -2.510  -0.000   0.000
    4    1    2    3        3       0.979   2.937   0.000  -3.916  -0.000   0.000
    5    1    2    3        3       0.979   2.937   0.000  -3.916  -0.000   0.000
    7    3    2    1        3      -0.444   3.833   0.728  -4.117  -0.000   0.000
    7    3    2    6        3       0.736   2.209   0.000  -2.946  -0.000   0.000

[ pairs ]
     3     4    1
     3     5    1
     4     6    1
     5     6    1
     6     7    1
`
Then, how to generate the corresponding .ff input files @ricalessandri .

**Reply by fgrunewald** [reply]
@YanChen32 the idea I believe is that you can use the script, which you currently cannot access, and that would extract the ff-files from the itp file you generated. The issue is that apparently we have not yet made the script publicly available. I'm sure @ricalessandri can add you to it. 

Otherwise you can also manually write the ff-files. You've already extracted the main repeat unit that is the `block`, so you can simply use that as one part of the input. Now you only need to write how `links` connect the blocks. We have written a [tutorial here](https://github.com/marrink-lab/polyply_1.0/wiki/Tutorial:-writing-.ff-input-files) for GROMOS but it works the same way for OPLS. Below I've written two links that should cover everything but pairs. For the terminal groups you repeat the process. 

```
; example link you need to fill in the values for the bonds and angles
[ link ]
resname "PVA"
[ bonds ]
C02  +C01  1
[ angles ]
C02 +C01 +C02  1
C01  C02 +C01  1
[ dihedrals ]
C01 C02 +C01 +C02 3 
; I think OPLS defines these dihedrals too but make sure to check
H04  C01 C02 +C01 3
H05  C01 C02 +C01 3
H06  C02 +C01 +C02 3
H06  C02 +C01 +H04 3
H06  C02 +C01 +H05 3
O03  C02 +C01 +C02 3
O03  C02 +C01 +H05 3
O03  C02 +C01 +H05 3
H07 O03 C02 +C01 3
; don't forget the pairs

[ link ]
resname "PVA"
[ dihedrals ]
C02 +C01 +C02 ++C01 3
```

**Reply by ricalessandri** [reply]
Hi @YanChen32 , so indeed the small script called `itp_to_ff.py` is not public yet. @fgrunewald and I will work to publish an open-beta version soon, at least we need to put down some decent documentation for it. In the meantime, I uploaded the correct (i.e., validated against oligomers produced by LigParGen) `PVA.ff` I produced with `itp_to_ff` [**via this PR**](https://github.com/marrink-lab/polyply_1.0/pull/324). Feel free to use it, just double-check that everything is in order.

Regarding the question about the difference in charges in termini, are you referring to -CH3 termini between PEO and PVA? If you see significant differences, you should go with different termini blocks, in my opinion. However, deciding how significant is significant, might be tricky.

Final note: as Fabian mentioned, PEO in [the library](https://github.com/marrink-lab/polyply_1.0/tree/master/polyply/data/oplsaaLigParGen) currently is only available OH-terminated. In case you want it -CH3-terminated, you'll need to define the CH3 block and it link to the PEO block. It should be relatively simple but let me know if you need help with it. In any case it may be good for us to have it in the library.

**Reply by YanChen32** [reply]
Thanks a lot for the help from @ricalessandri and @fgrunewald. I have already attempted manually write the ff-files, but it's a laborious process that is susceptible to errors and omissions. Hence, I am eagerly anticipating the open-source release of the script "itp_to_ff.py" as it will greatly expedite the implementation of OPLS-AA force fields in Polyply.

Regarding the PVA system, could you assist me in generating the ff-file with -OH termini? I'm prepared to test both systems with two distinct terminal groups.

**Reply by ricalessandri** [reply]
@YanChen32 I've added the OH-terminated PVA to that PR, see [the files](https://github.com/marrink-lab/polyply_1.0/pull/324/files).

**Comment by YanChen32** [comment]
Hi  @ricalessandri and @fgrunewald, 

I am writing to seek your guidance on some questions regarding the use of the build_file.bld for constructing PEO and PVA in confined spaces. I have encountered specific inquiries, and I appreciate any assistance you can provide. 

Firstly, I construct PEO500.itp and PVA68.itp using the following commands:

```
polyply gen_params -f PEO.oplsaa.LigParGen.ff -seq OHter:1 PEO:500 OHter:1 -o PEO500.itp -name PEO
``` 
and
```
polyply gen_params -f PVA_CH3ter.oplsaa.LigParGen.ff -seq CH3a:1 PVA:68 CH3b:1 -o PVA68.itp -name PVA
```

Next, I would like to confine the polymer model within a narrow slit space: [-25.0305,-5.81625, 1.75]->[25.0305,-5.81625,3.25]. Therefore, I have written the following bld file:

```
[ molecule ]
; molname
PEO 0 20
[ rectangle ]
; resname start stop  inside-out  x  y  z    a  b  c
 OHter      1      1     in       25.0305 5.81625 2.5  25.0305 5.81625 0.75
 PEO        2    501     in       25.0305 5.81625 2.5  25.0305 5.81625 0.75
 OHter    502    502     in       25.0305 5.81625 2.5  25.0305 5.81625 0.75

[ molecule ]
; molname
PVA 20 69
[ rectangle ]
; resname start stop  inside-out  x  y  z    a  b  c
 CH3a      1      1     in       25.0305 5.81625 2.5  25.0305 5.81625 0.75
 PVA       2     69     in       25.0305 5.81625 2.5  25.0305 5.81625 0.75
 CH3b      70    70     in       25.0305 5.81625 2.5  25.0305 5.81625 0.75
```
And the system file:
```
#include "oplsaa.ff/forcefield.itp"
#include "PEO500.itp"
#include "PVA68.itp"

[ system ]
; name
Melt of PEO and PVA

[ molecules ]
; name  number
PEO 20
PVA 49
```

After modeling with the following command:
```
polyply gen_coords -p system.top -b build_file.bld -o melt.gro -box 50.061 11.6325 5.0 -name melt
```

I observed that the polymer is not completely confined within the designated space. Here is an image illustrating the issue
<img width="800" alt="pic" src="https://github.com/marrink-lab/polyply_1.0/assets/54890252/2035baf1-a26f-4319-b154-cd164d111352">

I have spent a considerable amount of time trying to address this problem but haven't been successful. I would greatly appreciate your assistance in resolving this issue. Thank you for your time and support.

**Reply by fgrunewald** [reply]
Hi @YanChen32, Riccardo and myself are currently on conferences/vacation. I'll be back in the office Monday and try to solve the riddle. From a quick glance the build file looks alright. The free floating atoms are a little bit concerning and for sure not what should be happening. Is this structure already energy minimized and does generating a copolymer melt without confinement work?

**Reply by fgrunewald** [reply]
Hi @YanChen32, 

I found the issues here, but I must admit it was rather subtle. First your resid ranges are slightly off, due to the admittedly not ideal definition of the resid ranges within polyply. Essentially you miss the terminal residues, which leads to these residues being ignored for the confinement. Secondly, because your slit is so narrow (1.6nm) there are some 'apparent' outliers. Polyply only places residues with an accuracy of 1 step length, which in this case is ~0.43 nm. Thus all atoms should be confined within a slit of 1.6nm +/- 0.43 nm. Usually the step length is much much smaller than the confinement, which is why it normally does not cause any problems. 

If I rerun the coordinate generation with the build file below and keep in mind that there is a tolerance of 0.43nm, then all residues are placed within the desired slit. Furthermore, to pull all the atoms into the desired slit I would recommend adding a [flat-bottomed position restraint](https://manual.gromacs.org/current/reference-manual/functions/restraints.html), which adds a harmonic force to atoms when they are outside the slit z dimension. In that way you pull the few atoms slightly placed outside the range inside.

Let me know if this solution works for you!

```
[ molecule ]
; molname
PEO 0 20
[ rectangle ]
; resname start stop  inside-out  x  y  z    a  b  c
 OHter      1      2     in       25.0305 5.81625 2.5  25.0305 5.81625 0.75
 PEO        2    501     in       25.0305 5.81625 2.5  25.0305 5.81625 0.75
 OHter    502    503     in       25.0305 5.81625 2.5  25.0305 5.81625 0.75

[ molecule ]
; molname
PVA 20 69
[ rectangle ]
; resname start stop  inside-out  x  y  z    a  b  c
 CH3a      1      2     in       25.0305 5.81625 2.5  25.0305 5.81625 0.75
 PVA       2     69     in       25.0305 5.81625 2.5  25.0305 5.81625 0.75
 CH3b      70    71     in       25.0305 5.81625 2.5  25.0305 5.81625 0.75
```

**Reply by YanChen32** [reply]
Thank you very much. You have essentially solved my problem completely. It is true that the residue ranges in the bld file are incorrect for the terminal residues. I must admit that I still don't fully understand the writing method used for specifying the residue range. If it is supposed to follow the definition [start, stop), why isn't there a standardized form?

```
 CH3a      1      2     in       25.0305 5.81625 2.5  25.0305 5.81625 0.75
 PVA       2     70     in       25.0305 5.81625 2.5  25.0305 5.81625 0.75
 CH3b      70    71     in       25.0305 5.81625 2.5  25.0305 5.81625 0.75
```
In my test, this format also works well.
![image](https://github.com/marrink-lab/polyply_1.0/assets/54890252/1147cec5-a379-4741-bfec-bfd92f3abbde)
The presence of atoms outside the specified residue ranges should be solely attributed to accuracy issues (0.43nm). 

To confine all the atoms within the desired slit, I employ the Wall command in Gromacs. I will heed your suggestion and attempt to incorporate a flat-bottomed position restraint.

**Reply by YanChen32** [reply]
Hi  @fgrunewald,

When I change the bld file to
```
[ molecule ]
; molname
PEO 0 20
[ rectangle ]
; resname start stop  inside-out  x  y  z    a  b  c
 OHter      1      2     in       25.0305 5.81625 2.5  25.0305 5.81625 0.75
 PEO        2    502     in       25.0305 5.81625 2.5  25.0305 5.81625 0.75
 OHter    502    503     in       25.0305 5.81625 2.5  25.0305 5.81625 0.75

[ molecule ]
; molname
PVA 20 69
[ rectangle ]
; resname start stop  inside-out  x  y  z    a  b  c
 CH3a      1      2     in       25.0305 5.81625 2.5  25.0305 5.81625 0.75
 PVA       2     70     in       25.0305 5.81625 2.5  25.0305 5.81625 0.75
 CH3b      70    71     in       25.0305 5.81625 2.5  25.0305 5.81625 0.75

```
Perfect confinement of polymer can be obtained without the accuracy issues (0.43nm).
![image](https://github.com/marrink-lab/polyply_1.0/assets/54890252/db953e95-0bb7-4b29-88ce-3dbf0bcc6bf6)

**Reply by YanChen32** [reply]
Hi @fgrunewald,

Do you have any hints on how to incorporate a flat-bottomed position restraint in this system? I have noticed that there are very few tutorials available provided by Gromacs.

For instance, if I configure the following restraints in the ITP file:
```
[ position_restraints ]
; i funct g r k
1 2 5 1.5 1000
```
The ITP will apply a force whenever any molecule is below [COM(z) - 0.75] nm or above [COM(z) + 0.75] nm, where COM(z) represents the center of mass of the confined molecules.

In my case, I would like to confine the polymer within the range [2.5-0.75, 2.5+0.75]. Could you please advise on how to set up the restraint part in the ITP file?

Thank you for your continued support and guidance in answering my questions and clearing up any confusion!

**Reply by fgrunewald** [reply]
You're almost there, but the first index refers to a particular atom in your molecule not the molecule in general. To stay with the example the position restraints are applied to atom 1 of your molecule, when it is a distance greater than 1.5 away from the reference position. Now the reference position has to be defined in a reference .gro file that you need to give to GROMACS. So what you want to do is create a gro file where the z position of all your polymer atoms is 2.5nm that is your reference file. Then you add the position restraints to the itp file providing 0.75 as r parameters as shown below:

```
[ position_restraints ]
; i funct g r k
1 2 5 0.75 1000
2 2 5 0.75 1000
3 2 5 0.75 1000
.
.
.
``` 
After grompping you want to use mdrun as:
```
gmx mdrun -s topol.tpr -r reference.gro 
```

**Reply by YanChen32** [reply]
Thank you for your prompt response.

I have reviewed the setup of the flat-bottomed position restraint as you suggested. Specifically, the position restraint ITP file for PVA68.itp is named posrePVA.itp and contains the following information:
```
[ position_restraints ]
;  i funct      g      r          k
   1    2        5       0.8       100000
   2    2        5       0.8       100000
... ...
  68    2        5       0.8       100000

```
For PEO500.itp, we have set the posrePVA.itp which reads:
```
[ position_restraints ]
;  i funct      g      r          k
   1    2        5       0.8       100000
   2    2        5       0.8       100000
... ...
500   2        5       0.8       100000

```
In the system.top file, we have defined the position restraint as follows:
```
#include "oplsaa.ff/forcefield.itp"
#include "PEO500.itp"
#ifdef POSRESPEO
#include "posrePEO.itp"
#endif

#include "PVA68.itp"
#ifdef POSRESPVA
#include "posrePVA.itp"
#endif

[ system ]
; name
Melt of PEO and PVA

[ molecules ]
; name  number
PEO 20
PVA 49

```

In the MDP file, I have defined the position restraint using the following line:
```
define                   = -DPOSRESPVA -DPOSRESPEO
```
To obtain the reference.gro file, I followed your guidance and modified the z-coordinate of melt.gro by 2.5. Subsequently, I performed energy minimization using the following commands:
```
gmx grompp -f min.mdp -c melt.gro -p system.top -o min.tpr -r reference.gro  -maxwarn 10
gmx mdrun -v -deffnm min -s min.tpr -ntmpi 4 -ntomp 12
```
However, the resulting min.gro file does not fully confine the system within the ±8.0 space. I would greatly appreciate it if you could assist me in identifying any potential issues with the aforementioned settings.  I look forward to your prompt response.

**Reply by fgrunewald** [reply]
It looks like it should work. Can you share part of your reference gro file?

**Reply by YanChen32** [reply]
Something like this，
```
93916
    1OHter  OA1    1   8.230   3.182   2.5
    1OHter   C2    2   8.183   3.203   2.5
    1OHter  HA3    3   8.221   3.172   2.5
    1OHter   H4    4   8.185   3.196   2.5
    1OHter   H5    5   8.182   3.246   2.5
    2PEO    C01    6   7.846   3.299   2.5
    2PEO    O02    7   7.827   3.313   2.5
    2PEO    C03    8   7.861   3.293   2.5
... ...
   69PVA    H0541342  12.678   2.281   2.5
   69PVA    H0641343  12.732   2.205   2.5
   69PVA    H0741344  12.651   2.178   2.5
   70CH3b    C141345  12.957   2.164   2.5
   70CH3b    H141346  12.952   2.122   2.5
   70CH3b    H241347  12.998   2.162   2.5
   70CH3b    H341348  12.958   2.196   2.5
50.061 11.6325 5.0
```
The upper and lower bounds of the optimized strcutrue are approximately ± 11
![image](https://github.com/marrink-lab/polyply_1.0/assets/54890252/575b06c5-a712-484e-8134-013e0b05d50d)

**Reply by fgrunewald** [reply]
@YanChen32 honestly I don't exactly know. The fact that it remains a slab indicates that the flat-bottom potential does act. Otherwise it wouldn't remain a slab. My guess is that the density of the mixture is not compatible with the dimensions. That is you try to squeeze too much matter into this narrow slit. Ultimately, the LJ potential against which you work when squeezing an incompressible material will always win over a harmonic potential even if the force-constant is really high.

You could use gmx density to measure the density along z and check if that is approximately the amorphous density averaged from pure PVA and PEO. If the value is close it simply means you have too many polymer chains in the slit.

**Reply by YanChen32** [reply]
Thank you so much! You were absolutely correct in identifying the root cause of the problem: the density setting was too high. I was able to successfully resolve the issue by simply increasing the thickness of the slit.

**Reply by fgrunewald** [reply]
Perfect! If you don't have any further questions please feel free to mark this discussion as resolved
